src/libsyntax/parse/parser.rs

```diff
@@ -2700,6 +2697,99 @@ impl<'a> Parser<'a> {
+                // https://github.com/rust-lang/rust/issues/60610
+                self.expect(&token::Not)?;
+                self.expect(&token::OpenDelim(token::Paren))?;
+                let expr = self.parse_expr().map_err(|mut err| {
+                    err.span_label(
+                        await_sp,
+                        "while parsing this await macro call",
+                    );
+                    err
+                })?;
+                self.expect(&token::CloseDelim(token::Paren))?;
+                (expr.span, ExprKind::Await(ast::AwaitOrigin::MacroLike, expr))
+            }
+            token::Question => {
+                // Handle `await? <expr>`
+                self.bump(); // `?`
+                let expr = self.parse_expr().map_err(|mut err| {
+                    err.span_label(
+                        await_sp,
+                        "while parsing this incorrect await statement",
+                    );
+                    err
+                })?;
+                let sp = lo.to(expr.span);
+                let expr_str = self.sess.source_map().span_to_snippet(expr.span)
+                    .unwrap_or_else(|_| pprust::expr_to_string(&expr));
+                let expr = self.mk_expr(
+                    sp,
+                    ExprKind::Await(ast::AwaitOrigin::FieldLike, expr),
+                    ThinVec::new(),
+                );
+                let mut err = self.struct_span_err(
+                    sp,
+                    "incorrect use of `await`",
+                );
+                err.span_suggestion(
+                    sp,
+                    "`await` is not a statement",
+                    format!("{}.await?", expr_str),
+                    Applicability::MachineApplicable,
+                );
+                err.emit();
+                (sp, ExprKind::Try(expr))
+            }
+            ref t => {
+                // Handle `await <expr>`
+                let expr = if t == &token::OpenDelim(token::Brace) {
+                    // Handle `await { <expr> }`
+                    // this needs to be handled separatedly from the next arm to avoid
+                    // interpreting `await { <expr> }?` as `<expr>?.await`
+                    self.parse_block_expr(
+                        None,
+                        self.span,
+                        BlockCheckMode::Default,
+                        ThinVec::new(),
+                    )
+                } else {
+                    self.parse_expr()
+                }.map_err(|mut err| {
+                    err.span_label(
+                        await_sp,
+                        "while parsing this incorrect await statement",
+                    );
+                    err
+                })?;
+                let expr_str = self.sess.source_map().span_to_snippet(expr.span)
+                    .unwrap_or_else(|_| pprust::expr_to_string(&expr));
```

Comment:
This bit is duplicated from above. I have a feeling this pattern occurs frequently so it would be good to consider a cleanup after this PR (but deduplicate here for now).